server/pulp/plugins/rsync/publish.py

```diff
@@ -0,0 +1,522 @@
+        if self.delete:
+            args.append("--delete")
+        if self.links:
+            args.append("--links")
+        else:
+            args.append("--copy-links")
+        args.append(source_prefix)
+        args.append(self.make_destination(dest_prefix))
+        return args
+
+    def rsync(self):
+        """
+        This method formulates the rsync command based on parameters passed in to the __init__ and
+        then executes it.
+
+        :return: (boolean indicating success or failure, stdout + stderr)
+        :rtype: tuple
+        """
+        if not self.file_list:
+            return (True, "Nothing to sync")
+        output = ""
+        tmp_path1 = os.path.join(self.get_working_dir(), str(uuid.uuid4()))
+        open(tmp_path1, 'w').write("\n".join(sorted(self.file_list)))
+        # copy files here, not symlinks
+        (is_successful, this_output) = self.remote_mkdir(self.dest_prefix)
+        if not is_successful:
+            _logger.error("Cannot create directory %s: %s" % (self.dest_prefix, this_output))
+            return (is_successful, this_output)
+        output += this_output
+        rsync_args = self.make_rsync_args(tmp_path1, str(self.src_prefix),
+                                          self.dest_prefix, self.exclude)
+        (is_successful, this_output) = self.call(rsync_args)
+        _logger.info(this_output)
+        if not is_successful:
+            _logger.error(this_output)
+            return (is_successful, this_output)
+        output += this_output
+        return (is_successful, output)
+
+    def process_main(self):
+        (successful, output) = self.rsync()
+        if not successful:
+            raise PulpCodedException(message=output)
+
+
+class Publisher(PublishStep):
+    """
+    RSync publisher class that provides the common code for publishing to remote server. Each
+    plugin needs to extend this class and add necesary logic to the _add_necesary_steps method.
+    """
+    def __init__(self, repo, publish_conduit, config, distributor_type):
+        """
+        :param repo: Pulp managed Yum repository
+        :type  repo: pulp.plugins.model.Repository
+        :param publish_conduit: Conduit providing access to relative Pulp functionality
+        :type  publish_conduit: pulp.plugins.conduits.repo_publish.RepoPublishConduit
+        :param config: Pulp configuration for the distributor
+        :type  config: pulp.plugins.config.PluginCallConfiguration
+        """
+
+        super(Publisher, self).__init__("Repository publish", repo,
+                                        publish_conduit, config,
+                                        distributor_type=distributor_type,
+                                        working_dir="")
+
+        distributor = Distributor.objects.get_or_404(repo_id=self.repo.id,
+                                                     distributor_id=publish_conduit.distributor_id)
+        self.last_published = distributor["last_publish"]
```

Comment:
Since it's a proper object, I think we'd prefer accessing with normal attribute notation. Minor suggestion, but it makes it more clear what we're dealing with.